daemon/pool.go

```diff
@@ -73,123 +100,445 @@ func NewDriverPool(factory FactoryFunction) *DriverPool {
 	dp := &DriverPool{
 		ScalingPolicy: DefaultScalingPolicy(),
 		Logger:        logrus.New(),
-
-		factory: factory,
-		close:   make(chan struct{}),
-		queue:   newDriverQueue(),
+		factory:       factory,
 	}
-	dp.drivers.byID = make(map[string]Driver)
 	return dp
 }
 
 // Start stats the driver pool.
 func (dp *DriverPool) Start(ctx context.Context) error {
-	if dp.running {
+	if dp.stop != nil {
 		return ErrPoolRunning.New()
-	} else if dp.closed {
+	}
+	select {
+	case <-dp.stopped:
 		return ErrPoolClosed.New()
+	default:
+	}
+
+	dp.stop = make(chan struct{})
+	dp.stopped = make(chan struct{})
+	dp.spawn = make(chan struct{})
+	dp.spawnErr = make(chan error)
+	dp.rescale = make(chan struct{}, 1)
+	dp.get = make(chan driverRequest)
+	dp.put = make(chan Driver)
+	dp.drivers.idle = make(map[Driver]struct{})
+	dp.drivers.all = make(map[Driver]struct{})
+
+	dp.target.Set(1)
+
+	dp.wg.Add(3)
+	go dp.runSpawn()
+	go dp.runPolicy()
+	go dp.manageDrivers()
+
+	// wait for a single instance to come up
+	d, err := dp.getDriver(ctx)
+	if err != nil {
+		_ = dp.Stop()
+		return err
 	}
-	target := dp.ScalingPolicy.Scale(0, 0)
-	if err := dp.setInstances(ctx, target); err != nil {
-		_ = dp.setInstances(context.Background(), 0)
+	if err := dp.putDriver(d); err != nil {
 		return err
 	}
-	dp.running = true
-
-	go dp.scaling()
 	return nil
 }
 
-// setInstances changes the number of running driver instances. Instances
-// will be started or stopped as necessary to satisfy the new instance count.
-// It blocks until the all required instances are started or stopped.
-func (dp *DriverPool) setInstances(ctx context.Context, target int) error {
-	if target < 0 {
-		return ErrNegativeInstances.New()
+// runPolicy goroutine re-evaluates the scaling policy on a regular time interval and sets
+// a target number of instances. The scaling itself will be performed by the manager goroutine.
+func (dp *DriverPool) runPolicy() {
+	defer dp.wg.Done()
+	ticker := time.NewTicker(time.Millisecond * 100)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-dp.stop:
+			return
+		case <-ticker.C:
+		}
+		total := dp.running.Value()
+		load := dp.requests.Value()
+		dp.drivers.RLock()
+		idle := len(dp.drivers.idle)
+		dp.drivers.RUnlock()
+
+		target := dp.ScalingPolicy.Scale(total, load-idle)
+		if target < 1 {
+			target = 1 // there should be always at least 1 instance
+		}
+		old := dp.target.Set(target)
+		if old != target {
+			// optionally signal to the manager goroutine
+			select {
+			case dp.rescale <- struct{}{}:
+			default:
+			}
+		}
 	}
+}
 
-	dn := target - dp.stats.instances.Value()
-	if dn == 0 {
-		return nil
-	}
-	if dn > 0 {
-		return dp.add(ctx, dn)
+// spawnOne starts a new driver instance. It will keep trying to run it in case of a failure.
+func (dp *DriverPool) spawnOne() {
+	dp.spawning.Add(1)
+	defer dp.spawning.Add(-1)
+
+	// TODO(dennwc): use exponential backoff instead?
```

Comment:
I cannot think of any other good way to do it without a backoff.

If there is a fixed upper bound on retries, it may quickly spend all those retry attempts and kill bblfshd because of the panic.

In case we increase the delay between retries, it may be too slow to retry and requests will be waiting for this delay even on single failure.

So I guess I should add the backoff and the upper bound at the same time. Maybe in the next PR? Or should I do it here?

Reply:
I think it's fine to defer it.